### ISSUE TITLE
Avoiding merge conflicts for .gitrepo files for repository "owners"

### ISSUE THREAD

**Issue by tclune** [ingydotnet/git-subrepo#305]
A bit of background: In my organization, it is essential that most
developers be blissfully unaware of the use of git-subrepo.  I think
these are "owners" in the nomenclature introduced on the main git-repo
documentation page.  We expect a small number of gatekeepers to have
greater fluency with git and to manage any subrepo interactions.

My concern is that it seems relatively easy to have situations in
which the owners may encounter a conflict in the .gitrepo file
when merging 2 different branches in the main repo.  (I have a simple
synthetic script that demonstrates this, but I assume the issue is
unsurprising to those that are reading this question.)

So the big question is whether best practices exist for using
git-subrepo to eliminate or at least minimize occurrences when owners
will encounter merge conflicts in the .gitrepo file.  From my simple
script it seems that we should only (directly) use git-subrepo from a
single branch in our main repository.    Conflicts should then only arise in 
the branch that the gatekeepers manage.

A related question, that ultimately led to the question above, is
whether it can make sense to have 2 dedicated branches in the main
repo that track 2 long-lived branches in the subrepo.  My original,
probably naive, thought was to have a dev and release branch in the
subrepo.  The main repository would then have corresponding branches
that were used to track these.  But it would seem that subsequent
merges into the other long-lived branches in our main repo (dev,
integration, release) would effectively make a mockery of this
approach.  But maybe I'm missing some simple element of the strategy
that makes this something sensible to do.

Sorry for such naive questions.  My git-experience to date is
generally limited to small projects with 1-2 developers.

**Comment by grimmySwe** [comment]
@tclune Good questions!

Note that .gitrepo file only changes if someone uses a subrepo command. And by your description you have designated people doing this. If you have super-users that perform the subrepo commands they should also be aware of the implications.

I would recommend simplicity, don't create new branches until you really need them. In worst case you can even keep some changes in main repo only. Example is if you have a release branch that you don't want to move to much.

For me a narrow git structure feels best, keeping things close to master at all the time. Using short lived feature branches in the main repo. If you branch off into a release branch and you need support for this in the subrepo, create a branch at that time.

Was that any helpful? Please let me know if you need something more.

**Comment by tclune** [comment]
@grimmySwe 

Yes - your comments are helpful.    I like the idea of keeping some changes only in the main repo - at least for a time.    Perhaps then we could then treat the subrepo more as an external product and the subdir as having a vendor branch.       I'll try do do some gedanken experiments to see if this reduces/avoids the issue I raised.

Regarding a narrow git structure:   No argument from me.  However, our situation is complex and somewhat atypical.     I am trying to establish firm repo policies before the org converts from CVS.  (Yes - still using CVS due to over reliance on CVS modules.)    Unfortunately, this is a large, complex scientific model with very long time scales for validating scientific contributions.  (May have to simulate decades of weather/climate to determine if the change improves the model fidelity; tweak parameters and then repeat.)  During these validation periods, the researchers generally will not risk merging in other changes that might introduce minor diffs in the results.    Even if we can keep that work from polluting the main repository with relatively long-lived branches, there will be limits on how close to master most of the researchers can actually work.    

In my dreams, I will push things towards continuous integration.       In practice, I fear we will average at least one long-lived branch per researcher, but am hoping we can mostly keep that out of the main/shared repository.    But I will consider the git conversion a win if we can even get a typical researcher to commit their changes locally on a regular basis.     Our core team can then at least routinely integrate their changes and perform various types of regression tests.     

Based upon my experience "upgrading" a similar scientific organization, a major challenge will be to get the researchers to start their next project from the main dev/integration/release branch rather than from their increasingly obsolete branch -- no matter that we have absorbed their changes.

 On the model infrastructure side, I'm more confident that my team can approximate continuous integration.    Our changes can generally be validated with unit tests and our algorithm changes don't usually introduce round-off differences that otherwise confuse system-level regression tests.   (Chaotic dynamics are a terrible complication.)

Cheers.

**Comment by dhuantes** [comment]
@tclune 
I work with a research organization as well and we also used CVS prior to our move to git.  Unfortunately we used submodules before moving to subrepo.  So you are already ahead of the curve...IMHO...:-D...  Submodules are far closer to CVS modules but generally are a bad choice unless all of your developers are going to embrace git completely.  In git-subrepo terms if you're using submodules everyone has to be a "collaborator".  One thing to remember is that branching was very expensive in CVS so we rarely branched.  With git it's the opposite.  I would encourage you to look into GitLab (https://about.gitlab.com/)  for  your research organization.  It's just like GitHub but it's open source and you can host it on your own private/corporate server.  So in some of the cases you describe instead of having a branch that exists forever for each of your researchers you could choose to just have each researcher fork the entire repository for that group/team.   GitLab has continuous integration built in out of the box so you could work towards that on the main repo so that when you recieve merge requests from any of the forked repositories you can run their changes through your CI scripts before they're accepted and actually merged.  Good Luck!

**Comment by tclune** [comment]
We use CVS modules to _completely_ rearrange the directory structure for different configurations.   We have explored various git approaches to maintaining this approach, but our current plan is to force the researchers to work with the "global" module.     Even if we embraced git submodules to attempt to preserve the status quo, we would need heavily nested submodules to provide relevant views of the repro and it would quickly become untenable.

If you are interested, I did go a long way towards a solution in which we used symlinks to create a directory tree in which each top level subdirectory corresponded to a CVS module.    Aside from a few subtleties that I had to resolve in the build system, the big problem was the fact that "cd .." would not always do what one would expect.    Ultimately we decided to abandon this, but only after a considerable amount of work on my part.

Instead we will convert the "global" CVS module as the initial commit in the git repository and use git sparsetree to limit the view of the directory tree a bit.    This can't do wholesale rearranging of the directory structure but for the most commonly used modules it will look familiar to the researchers.
 I nonetheless expect this will be the largest pushback during the transition.    Just have to keep reminding them about what the real functionality they need and how git provides that - just by different means.       I have my own preferred long term solution in which we use a flatter directory structure, but I've not yet had any converts in the relevant stakeholders.

NASA currently licenses an external and an internal GitHub.    Our plan is to use the internal GitHub while setting everything up and then switch to the public GitHub when we go live.   (The internal one is not accessible to foreign nationals, much less various external collaborators.)    We are not opposed to GitLab, but lacking direct experience we are unable to evaluate the pros and cons of using the services that are already on the table vs. the difficulty of deploying a new solution that is still accessible to our external collaborators.    You say that it is "just like GitHub" which would imply there is no advantage given that our org already has GitHub.   But I'm happy to learn about relevant differences ...
